lib/rules/number-max-precision/index.js

```diff
@@ -72,47 +89,123 @@ const rule = (primary, secondaryOptions) => {
+				if (isFunctionNode(componentValue) || isSimpleBlockNode(componentValue)) {
+					componentValue.walk((entry) => {
+						if (!entry.state) return;
+
+						if (entry.state.ignored) return;
+
+						walker(node, getIndex, entry.node, entry.state);
+					}, state);
 				}
+			});
+		}
+
+		/**
+		 * @template {import('postcss').AtRule | import('postcss').Declaration} T
+		 * @param {T} node
+		 * @param {(node: T) => number} getIndex
+		 * @param {import('@csstools/css-parser-algorithms').ComponentValue} componentValue
+		 * @param {{ ignored: boolean, precision: number }} state
+		 */
+		function walker(node, getIndex, componentValue, state) {
+			if (isFunctionNode(componentValue)) {
+				const name = componentValue.getName().toLowerCase();
+
+				if (name === 'url') {
+					state.ignored = true;
 
-				// Ignore strings, comments, etc
-				if (valueNode.type !== 'word') {
 					return;
 				}
 
-				const match = /\d*\.(\d+)/.exec(valueNode.value);
+				state.precision = precisionInsideFunction(name, state.precision);
+
+				return;
+			}
+
+			if (!isTokenNode(componentValue)) {
+				return;
+			}
+
+			const [tokenType, raw, startIndex, endIndex, parsedValue] = componentValue.value;
+
+			if (
+				tokenType !== TokenType.Number &&
+				tokenType !== TokenType.Dimension &&
+				tokenType !== TokenType.Percentage
+			) {
+				return;
+			}
+
+			let unitStringLength = 0;
+
+			if (tokenType === TokenType.Dimension) {
+				const unit = parsedValue.unit;
+
+				unitStringLength = unit.length;
 
-				if (match == null || match[0] == null || match[1] == null) {
+				if (optionsMatches(secondaryOptions, 'ignoreUnits', unit)) {
 					return;
 				}
+			} else if (tokenType === TokenType.Percentage) {
+				unitStringLength = 1;
 
-				if (match[1].length <= primary) {
+				if (optionsMatches(secondaryOptions, 'ignoreUnits', '%')) {
 					return;
 				}
+			}
+
+			const match = /\d*\.(\d+)/.exec(raw);
+
+			if (match == null || match[0] == null || match[1] == null) {
+				return;
+			}
+
+			if (match[1].length <= state.precision) {
+				return;
+			}
 
-				const baseIndex = isAtRule(node) ? atRuleParamIndex(node) : declarationValueIndex(node);
-				const actual = Number.parseFloat(match[0]);
-
-				report({
-					result,
-					ruleName,
-					node,
-					index: baseIndex + valueNode.sourceIndex + match.index,
-					word: actual.toString(),
-					message: messages.expected,
-					messageArgs: [actual, actual.toFixed(primary)],
-				});
+			const nodeIndex = getIndex(node);
+
+			report({
+				result,
+				ruleName,
+				node,
+				index: nodeIndex + startIndex,
+				endIndex: nodeIndex + (endIndex + 1) - unitStringLength,
+				message: messages.expected,
+				messageArgs: [parsedValue.value, parsedValue.value.toFixed(state.precision)],
 			});
 		}
+
+		/**
+		 * @param {string} functionName
+		 * @param {number} currentPrecision
+		 * @returns {number}
+		 */
+		function precisionInsideFunction(functionName, currentPrecision) {
+			{
```

Comment:
I've added a block here because otherwise `precision` is declared twice.

I don't have a strong preference here.
Changing the names would also work.

Reply:
Hum, how about the name `precisionForFunction` instead of `precision`?